### DISCUSSION TITLE
What is best practice to initialise a NotifierProvider with a parameter from a widget

### DISCUSSION THREAD

**Discussion by aidanmack** [rrousselGit/riverpod#3720] [Q&A] [answered]
Im working with some legacy code which does not entirely use riverpod and so I need to initialise my NotifierProvider when the widget starts watching it.

```
  Widget build(BuildContext context, WidgetRef ref) {


    final state = ref.watch(cardServicesProvider);
    ref.read(cardServicesProvider.notifier).buildCardServicesForQuickView(cardId);
```

I get these suggestions:

_For example, maybe you could update your provider inside a button's
  onPressed instead_

and

_Future(() {...})_

Is using a future the only way I can do thi?

I've also seen work arounds making the widget stateful and using 

  @override
  void initState() {
    super.initState();
    //PostFrameCallback((_) 
}

**Comment by snapsl** [comment, marked as answer]
I would suggest using a family provider to pass the card id.
``` dart
final state = ref.watch(cardServicesProvider(cardId));
```

**Reply by aidanmack** [reply]
Yeah I thought that and and in this context I think I might.
But I'm always a bit on the fence about using family as if further into development something else wants to "watch" changes on that provider, it has to supply the parameter. Sometimes you just want to watch a families providers current state with out initialising it with parameter. Ive never seen a solution to that.
